packages/schematics/angular/utility/validation.ts

```diff
@@ -11,9 +11,16 @@ import { SchematicsException } from '@angular-devkit/schematics';
 // Must start with a letter, and must contain only alphanumeric characters or dashes.
 // When adding a dash the segment after the dash must also start with a letter.
 export const htmlSelectorRe = /^[a-zA-Z][.0-9a-zA-Z]*(:?-[a-zA-Z][.0-9a-zA-Z]*)*$/;
+const classNameRegExp = /^[a-zA-Z_]\w*$/;
```

Comment:
`$` is allowed as well as some unicode characters.
There's actually a nice example of testing for a valid ECMAScript identifier in the regexp unicode property proposal: https://github.com/tc39/proposal-regexp-unicode-property-escapes/blob/fe6d07fad74cd0192d154966baa1e95e7cda78a1/README.md#other-examples